### PR TITLE
fixes #26198 - ensure host list does not require full page reload

### PR DESCRIPTION
**Pull request by ohadlevy**

### CODE REVIEW

File: webpack/assets/javascripts/hosts/tableCheckboxes.js [hunk 1]
```diff
@@ -0,0 +1,249 @@
+import { sprintf, n__, translate as __ } from '../react_app/common/I18n';
+
+// Array contains list of host ids
+$.cookieName = `_ForemanSelected${window.location.pathname.replace(/\//, '')}`;
```
Comment (ekohl) [inline webpack/assets/javascripts/hosts/tableCheckboxes.js]: Don't you need to import `$`?

Comment (ohadlevy) [reply webpack/assets/javascripts/hosts/tableCheckboxes.js]: thanks, updated, and also moved the deprecated jquery.cookie next to the code.

**Comment by ohadlevy** [comment]
I've added depreciation functions as it seems discovery / scap uses it.

File: webpack/assets/javascripts/bundle.js [hunk 2]
```diff
@@ -30,6 +29,7 @@ import compute from './foreman_compute_resource';
 import componentRegistry from './react_app/components/componentRegistry';
 import i18n from './react_app/common/I18n';
 import * as foremanDocument from './react_app/common/document';
+import hosts from './foreman_hosts';
```
Comment (ezr-ondrej) [inline webpack/assets/javascripts/bundle.js:32]: Why not to import all of the dependencies, but just one? Looks bit weird now.

File: webpack/assets/javascripts/foreman_hosts.js [hunk 3]
```diff
@@ -1,4 +1,13 @@
 import { uniq, set } from 'lodash';
+import * as table from './hosts/tableCheckboxes';
+
+export default {
```
Comment (ezr-ondrej) [inline webpack/assets/javascripts/foreman_hosts.js:4]: Not sure if this is preferred way of exporting, maybe it is - it is stating at the beginning all the exports. but I would deleted the other exports than - why to export same stuff twice? :thinking:

File: webpack/assets/javascripts/hosts/tableCheckboxes.js [hunk 4]
```diff
@@ -0,0 +1,251 @@
+import $ from 'jquery';
+import { sprintf, n__, translate as __ } from '../react_app/common/I18n';
+
+require('jquery.cookie');
+
+// Array contains list of host ids
+$.cookieName = `_ForemanSelected${window.location.pathname.replace(/\//, '')}`;
+$.foremanSelectedHosts = readFromCookie();
```
Comment (ezr-ondrej) [inline webpack/assets/javascripts/hosts/tableCheckboxes.js]: I know, you are just moving it, not writing, but this assign seems especially fishy, maybe would be nice to rewrite it?
Do not understand the purpose, but assigning some property to the jquery doesn't sound like a good idea, I would suggest to put it at least in `window` in this PR.

**Review by ezr-ondrej — COMMENTED** [review]
Looks working, just the host selection preservation stopped working (the checkboxes are not staying checked after taking an action).

File: webpack/assets/javascripts/bundle.js (hunk 2 above)
Comment (ohadlevy) [reply webpack/assets/javascripts/bundle.js:32]: this allow to consume the functions both via tfm.hosts.. and as a simple import {x} from ...

I did it in order to not break code base, and its a pattern we already have (e.g. see compute resources).

**Comment by ohadlevy** [comment]
I've:
- rebased
- fixed the issue where content was no longer visible (after #6502)
- refactored $. variables
- moved foreman_url function to foremanTools
- used foremanUrl function
- fixed the issue with checkboxes not being persistent

File: webpack/assets/javascripts/hosts/tableCheckboxes.js [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+}
+
+export function buildRedirect(url) {
+  const data = getBulkParam();
+  const uri = `${url}?${$.param(data)}`;
+  window.location.replace(uri);
+}
+
+function paginationMetaData() {
+  const pagination = $('#pagination');
```
Comment (ohadlevy) [inline webpack/assets/javascripts/hosts/tableCheckboxes.js:232]: @glekner @amirfefer this seems to be broken after pagination refactoring, its outside the scope of this PR, so its mostly FYI as I've encountered it.

**Comment by ohadlevy** [comment]
hmm.. test failures do not seem related.

**Comment by ohadlevy** [comment]
[test foreman]

File: webpack/assets/javascripts/bundle.js (hunk 2 above)
Comment (ezr-ondrej) [reply webpack/assets/javascripts/bundle.js:32]: oh, ok, make sense, thanks for explanation :+1:

**Comment by ohadlevy** [comment]
rebased and removed the workaround for showing the content as it was fixed in #6545

**Comment by ohadlevy** [comment]
tests failures do not seem related, @tbrisker mind having a look? thanks.

**Comment by ohadlevy** [comment]
anything else needed for this PR? thanks.

**Review by ezr-ondrej — DISMISSED** [review]
Code looks good to me and what I could come up with to test is working!
Good job @ohadlevy! :+1:

**Comment by ares** [comment]
rebase needed now

**Comment by ohadlevy** [comment]
rebased.

**Comment by ohadlevy** [comment]
[test foreman]
[test katello]

**Comment by ohadlevy** [comment]
@tbrisker do you mind reviewing? I believe its all ready for a while now.

**Review by amirfefer — DISMISSED** [review]
Tested and works well (excluding the choose all host regression :( )
The hosts page is loading faster, and overall it's a great change 
Thanks @ohadlevy  :+1:

**Comment by ohadlevy** [comment]
anyone else can merge? perhaps @timogoebel ?

**Comment by ohadlevy** [comment]
test failures not related.

File: webpack/assets/javascripts/hosts/tableCheckboxes.js [hunk 6]
```diff
@@ -0,0 +1,252 @@
...
+// Array contains list of host ids
+const cookieName = `_ForemanSelected${window.location.pathname.replace(
+  /\//,
+  ''
+)}`;
+let foremanSelectedHosts = readFromCookie();
+
+// triggered by a host checkbox change
+export function hostChecked({ id, checked }) {
+  const multpleAlert = $('#multiple-alert');
```
Comment (timogoebel) [inline webpack/assets/javascripts/hosts/tableCheckboxes.js]: typo: should be `multipleAlert` (missing `i`)

File: webpack/assets/javascripts/hosts/tableCheckboxes.js (hunk 5 above)
Comment (timogoebel) [reply webpack/assets/javascripts/hosts/tableCheckboxes.js:232]: Do we have a redmine issue to fix this?

**Review by timogoebel — CHANGES_REQUESTED** [review]
@ohadlevy: I found one small typo, otherwise: 👍

File: webpack/assets/javascripts/hosts/tableCheckboxes.js (hunk 5 above)
Comment (ohadlevy) [reply webpack/assets/javascripts/hosts/tableCheckboxes.js:232]: https://projects.theforeman.org/issues/26494

**Comment by ohadlevy** [comment]
@timogoebel updated - thanks, test failures are unrelated.

**Review by timogoebel — APPROVED** [review]

**Comment by timogoebel** [comment]
Thanks, @ohadlevy.

**Pull request merged**